File: java/cours/src/session2/td/agregation/NaturelInductif.java
package session2.td.agregation;

public interface NaturelInductif extends EtatNaturelPur {
	
	
	@Override default
	public EtatNaturelPur creerZero() {
		return new ZeroInductif();
	}

	@Override default
	public EtatNaturelPur creerSuccesseur(EtatNaturelPur predecesseur) {
		return new SuccInductif(predecesseur);
	}
	
	@Override
	default EtatNaturelPur creerNatAvecValeur(int val) {
		if(val == 0)
			return this.creerZero();
		return this.creerSuccesseur(this.creerNatAvecValeur(val -1));
	}
}

File: typescript/session2/td/agregation/NaturelInductif.ts
import { EtatNaturelPur } from "./EtatNaturelPur";
import { SuccInductif } from "./SuccInductif";
import { ZeroInductif } from "./ZeroInductif";
import { heriterInterfacesConcretes } from "../../../heritageMultiple/heritageInterfacesconcretes";
import { NatInductif_defaut } from "../heritageMultiple/NatInductif";

export interface NaturelInductif extends EtatNaturelPur {
    creerZero() : EtatNaturelPur;
    creerSuccesseur(predecesseur : EtatNaturelPur) : EtatNaturelPur;
    creerNatAvecValeur(valeur : number) : EtatNaturelPur;
}

export abstract class NaturelInductif_defaut implements EtatNaturelPur{
    creerNatAvecValeur(valeur : number) : EtatNaturelPur {
		if (valeur === 0)
			return this.creerZero();
		return this.creerSuccesseur(this.creerNatAvecValeur(valeur - 1));
	}
	creerZero() : EtatNaturelPur {
		return new ZeroInductif();
	}
	creerSuccesseur(predecesseur : EtatNaturelPur ) : EtatNaturelPur {
		return new SuccInductif(predecesseur);
    }
    
    abstract creerNatAvecRepresentation(repDecimal: string): EtatNaturelPur;
    abstract estNul(): boolean;
    abstract predecesseur(): EtatNaturelPur;
    abstract chiffre(i: number): number;
    abstract taille(): number;
    abstract val(): number;
}

heriterInterfacesConcretes(NaturelInductif_defaut, [NatInductif_defaut])
